osquery/tables/system/windows/registry.h

```diff
@@ -0,0 +1,160 @@
+#include <string>
+#include <vector>
+// Used for converting Big Endian to Little Endian
+#include <WbemIdl.h>
+#include <Windows.h>
+#include <intrin.h>
+#include <osquery/tables.h>
+
+namespace osquery {
+namespace tables {
+
+// Windows registry helper macros for key and path sizes
+#define MAX_KEY_LENGTH 255
+#define MAX_VALUE_NAME 16383
+
+/// Takes a Windows FILETIME object and returns seconds since epoch
+static LONGLONG filetime_to_unixtime(FILETIME ft) {
+  LARGE_INTEGER date, adjust;
+  date.HighPart = ft.dwHighDateTime;
+  date.LowPart = ft.dwLowDateTime;
+  adjust.QuadPart = 11644473600000 * 10000;
+  date.QuadPart -= adjust.QuadPart;
+  return date.QuadPart / 10000000;
+};
+
+/// Microsoft helper function for getting the contents of a registry key
+std::vector<std::map<std::string, std::string>> QueryKey(HKEY hKey) {
+  TCHAR achClass[MAX_PATH] = TEXT("");
+  TCHAR achKey[MAX_KEY_LENGTH];
+  TCHAR achValue[MAX_VALUE_NAME];
+
+  DWORD cchValue = MAX_VALUE_NAME;
+  DWORD cbName;
+  DWORD cchClassName = MAX_PATH;
+  DWORD cSubKeys = 0;
+  DWORD cbMaxSubKey;
+  DWORD cchMaxClass;
+  DWORD cValues;
+  DWORD cchMaxValueName;
+  DWORD lpType;
+  DWORD cbMaxValueData;
+  DWORD cbSecurityDescriptor;
+  DWORD i, retCode;
+  size_t cnt = 0;
+
+  std::vector<std::string> multi_sz_strs;
+
+  FILETIME ftLastWriteTime;
+  std::vector<std::map<std::string, std::string>> res;
+
+  retCode = RegQueryInfoKey(hKey,
+                            achClass,
+                            &cchClassName,
+                            NULL,
+                            &cSubKeys,
+                            &cbMaxSubKey,
+                            &cchMaxClass,
+                            &cValues,
+                            &cchMaxValueName,
+                            &cbMaxValueData,
+                            &cbSecurityDescriptor,
+                            &ftLastWriteTime);
+
+  if (cSubKeys) {
+    for (i = 0; i < cSubKeys; i++) {
+      cbName = MAX_KEY_LENGTH;
+      retCode = RegEnumKeyEx(hKey, i, achKey, &cbName, NULL, NULL, NULL,
+                             &ftLastWriteTime);
+      if (retCode == ERROR_SUCCESS) {
+        std::map<std::string, std::string> m;
+        m["reg_type"] = "subkey";
+        m["index"] = std::to_string(i);
+        m["subkey"] = achKey;
+        m["last_write_time"] =
+            std::to_string(filetime_to_unixtime(ftLastWriteTime));
+        res.push_back(m);
+      }
+    }
+  }
+
+  BYTE* bpDataBuff = new BYTE[cbMaxValueData];
+  if (cValues) {
+
+    for (i = 0, retCode = ERROR_SUCCESS; i < cValues; i++) {
+      ZeroMemory(bpDataBuff, cbMaxValueData);
+      cchValue = MAX_VALUE_NAME;
+      achValue[0] = '\0';
+      retCode =
+          RegEnumValue(hKey, i, achValue, &cchValue, NULL, NULL, NULL, NULL);
+
+      if (retCode == ERROR_SUCCESS) {
+        DWORD lpData = cbMaxValueData;
+        LONG dwRes =
+            RegQueryValueEx(hKey, achValue, 0, &lpType, bpDataBuff, &lpData);
```

Comment:
A few of these structures are initialized at the top of the method, maybe move them down here for safety. If you construct them within the most specific scope you have a better guarantee of their state. E.g., if they are outside of a loop you need to understand if/when to reset them.